src/emd/rt_bse.F

```diff
@@ -0,0 +1,1420 @@
+      DO k = 1, 3
+         CALL cp_fm_scale_and_add(1.0_dp, rtbse_env%real_workspace(1)%matrix, &
+                                  kvec(k), rtbse_env%moments_field(k)%matrix)
+      END DO
+      ! enforce hermiticity of the effective Hamiltonian
+      CALL cp_fm_transpose(rtbse_env%real_workspace(1)%matrix, rtbse_env%real_workspace(2)%matrix)
+      CALL cp_fm_scale_and_add(0.5_dp, rtbse_env%real_workspace(1)%matrix, &
+                               0.5_dp, rtbse_env%real_workspace(2)%matrix)
+      ! Prepare the exponential/exponent for propagation
+      IF (rtbse_env%mat_exp_method == do_bch) THEN
+         ! Multiply by the S_inv matrix - in the classic ordering
+         CALL parallel_gemm("N", "N", rtbse_env%n_ao, rtbse_env%n_ao, rtbse_env%n_ao, &
+                            intensity, rtbse_env%S_inv_fm%matrix, rtbse_env%real_workspace(1)%matrix, &
+                            0.0_dp, rtbse_env%real_workspace(2)%matrix)
+         DO i = 1, rtbse_env%n_spin
+            ! Sets real part to zero
+            CALL cp_fm_to_cfm(msourcei=rtbse_env%real_workspace(2)%matrix, mtarget=rtbse_env%ham_effective(i)%matrix)
+         END DO
+      ELSE IF (rtbse_env%mat_exp_method == do_exact) THEN
+         DO i = 1, rtbse_env%n_spin
+            CALL cp_fm_to_cfm(msourcer=rtbse_env%real_workspace(1)%matrix, mtarget=rtbse_env%ham_workspace(i)%matrix)
+            CALL cp_cfm_gexp(rtbse_env%ham_workspace(i)%matrix, rtbse_env%S_cfm%matrix, rtbse_env%ham_effective(i)%matrix, &
+                             CMPLX(0.0, intensity, kind=dp), rtbse_env%rho_workspace)
+         END DO
+      END IF
+      ! Propagate the density by the effect of the delta pulse
+      CALL propagate_density(rtbse_env, rtbse_env%ham_effective, rtbse_env%rho, rtbse_env%rho_new)
+      metric = rho_metric(rtbse_env%rho_new, rtbse_env%rho, rtbse_env%n_spin)
+      IF (rtbse_env%unit_nr > 0) WRITE (rtbse_env%unit_nr, ('(A42,E38.8E3)')) " RTBSE| Metric difference after delta kick", metric
+      ! Copy the new density to the old density
+      DO i = 1, rtbse_env%n_spin
+         CALL cp_cfm_to_cfm(rtbse_env%rho_new(i)%matrix, rtbse_env%rho(i)%matrix)
+      END DO
+
+      CALL timestop(handle)
+   END SUBROUTINE apply_delta_pulse
+! **************************************************************************************************
+!> \brief Determines the metric for the density matrix, used for convergence criterion
+!> \param rho_new Array of new density matrices (one for each spin index)
+!> \param rho_old Array of old density matrices (one for each spin index)
+!> \param nspin Number of spin indices
+!> \param workspace_opt Optionally provide external workspace to save some allocation time
+! **************************************************************************************************
+   FUNCTION rho_metric(rho_new, rho_old, nspin, workspace_opt) RESULT(metric)
+      TYPE(cp_cfm_p_type), DIMENSION(:), INTENT(IN)      :: rho_new, &
```

Comment:
Do you really have to use arrays of cp_cfm_p_type, i.e. an array of pointers to cp_cfm_type or would it be too complicated to express what you want to achieve with arrays of cp_cfm_type? With cp_cfm_p_type, you introduce another layer of pointers which is in my experience not necessary. There are other occasions as well.

Reply:
I changed the type of all used matrix arrays in the code to be of `TYPE(matrix_type), DIMENSION(:), POINTER`, hopefully this is a more direct approach.